File: pepcoding/level up  pepcoding - june batch/1st_june/l001.java
import java.util.Scanner;
import java.util.ArrayList;

public class l001 {

    public static Scanner scn = new Scanner(System.in);

    public static void printEvenOdd(int a, int b)

    {
        if (a == b + 1) {
            return;
        }
        if (a % 2 == 0)

        {
            System.out.println(a);
        }

        printEvenOdd(a + 1, b);

        if (a % 2 != 0)

        {
            System.out.println(a);
        }
    }

    public static void set1() {

        int a = scn.nextInt();
        int b = scn.nextInt();
        printEvenOdd(a, b);

    }

    public static void set2() {

        int[] arr = { 1, 45, 67, 456, 345, 5, 454324, 34, 14, 1, 4312, 4 };
        // array with size 10 and default value as 0
        // in java except primitive data types everything is made on heap and primitive
        // data types are made on stack
        // * -- not sure -- * in cpp every thing is made on stack unless we say it to
        // make on heap
        display(arr, 0);

        System.out.println();
        System.out.println(firstIndex(arr, 0, 5));

        System.out.println(lastIndex(arr, 0, 4));

        System.out.println(allIndex(arr, 0, 67, 0));
        // this will return the address of the array and not the array itself

        int answer[] = allIndex(arr, 0, 67, 0);

        for (int i = 0; i < answer.length; i++) {
            System.out.print(answer[i] + " ");
        }

    }

    public static void display(int[] arr, int idx) {
        if (idx == arr.length) {
            return;
        }
        System.out.print(arr[idx] + " ");
        display(arr, idx + 1);
    }

    public static int firstIndex(int[] arr, int idx, int data) {

        if (idx == arr.length)
            return -1;

        if (arr[idx] == data)
            return idx;

        return firstIndex(arr, idx + 1, data);

    }

    public static int lastIndex(int[] arr, int idx, int data) {

        if (idx == arr.length)
            return -1;

        int ans = lastIndex(arr, idx + 1, data);

        if (ans != -1) {
            return ans;
        }

        else {
            return arr[idx] == data ? idx : -1;
        }

    }

    public static int[] allIndex(int[] arr, int idx, int data, int count) {
        // rec last se bhi chala sakte hain
        // yahan pe jo hai we are using post order type of revurssion
        if (idx == arr.length)
            return new int[count];

        if (arr[idx] == data) {
            int[] ans = allIndex(arr, idx + 1, data, count + 1);
            ans[count] = idx;
            return ans;

        } else {
            int ans[] = allIndex(arr, idx + 1, data, count);
            return ans;
        }

    }

    public static String[] words = { ":;/", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tu", "vwx", "yz", "&*%",
            "!@$" };

    public static void set3() {

        // System.out.println(nokiaKeyPad1("1010"));

        // System.out.println(nokiaKeyPad2("1010", ""));

        // System.out.println(printEncodings("10123", ""));
        // System.out.println("pulkit sood ");

    }

    public static ArrayList<String> nokiaKeyPad1(String str) {
        if (str.length() == 0) {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        char ch = str.charAt(0);
        ArrayList<String> ans = nokiaKeyPad1(str.substring(1));

        String word = words[ch - '0'];
        ArrayList<String> myans = new ArrayList<>();
        for (int i = 0; i < word.length(); i++) {
            for (String s : ans) {
                myans.add(word.charAt(i) + s);
            }
        }

        return myans;
    }

    public static int nokiaKeyPad2(String str, String ans)

    {

        if (str.length() == 0) {
            System.out.println(ans);
            return 1;
        }

        char ch = str.charAt(0);
        String newques = str.substring(1);
        String word = words[ch - '0'];

        int count = 0;

        for (int i = 0; i < word.length(); i++) {
            count += nokiaKeyPad2(newques, ans + word.charAt(i));
        }

        if (str.length() > 1) {
            char ch2 = str.charAt(1);
            int num = (ch - '0') * 10 + (ch2 - '0');

            if (num >= 10 && num <= 11) {
                word = words[num];
                for (int i = 0; i < word.length(); i++) {
                    count += nokiaKeyPad2(str.substring(2), ans + word.charAt(i));
                }
            }
        }
        return count;
    }

    public static void set4() {
        System.out.println(printEncodings("1423", ""));
        System.out.println("pulkit sood ");
    }

    public static int printEncodings(String str, String ans) {
        if (str.length() == 0) {
            System.out.println(ans);
            return 1;
        }

        int count = 0;
        char ch = str.charAt(0);

        if (ch == '0') {
            return 0;
        }

        String nstr = str.substring(1);
        count += printEncodings(nstr, ans + (char) (ch - '0' + 'a' - 1));

        if (str.length() > 1) {
            char ch2 = str.charAt(1);
            int num = (ch - '0') * 10 + (ch2 - '0');

            if (num >= 10 && num <= 25) {
                nstr = str.substring(2);
                count += printEncodings(nstr, ans + (char) (num + 'a' - 1));
            }
        }

        return count;
    }

    public static void solve() {

        set4();

    }

    public static void main(String[] args) {
        solve();
    }
}

File: pepcoding/level up  pepcoding - june batch/1st_june/l001.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

// void set1()
// {
//       int a,b;
//      cin>>a>>b;
//       printIncDec(a,b);

//     int a, b;
//     cin >> a >> b;

//     cout << powerSqrt(a, b) << endl;
// }
// void printIncDec(int a, int b)
// {
//     if (a == b + 1)
//     {
//         return;
//     }
//     cout << a << " hi " << endl;
//     printIncDec(a + 1, b);
//     cout << a << " bye " << endl;
// }
// int fact(int n)
// {
//     // if (n< = 1 ) return 1 ;
//     // return fact (n-1 ) * n ;
//     return n <= 1 ? 1 : fact(n - 1) * n;
// }

// int power(int a, int b)
// {
//     return b == 0 ? 1 : power(a, b - 1) * a;
// }
// int powerSqrt(int a, int b)
// {
//     // basecase

//     if (b == 0)
//         return 1;

//     int ans = powerSqrt(a, b / 2);
//     ans *= ans;
//     return b % 2 == 0 ? ans : ans * a;
// }

// void display(vector<int> &arr, int idx) // & hamesha lagana hai during passing vector

// {
//     if (idx == arr.size())
//         return;

//     cout << arr[idx] << " ";
//     display(arr, idx + 1);
// }

// int maximum(vector<int> &arr, int idx)
// {
//     if (idx == arr.size() - 1)
//         return arr[idx];

//     return max(maximum(arr, idx + 1), arr[idx]);
//     // in java
//     // return Math.max(smallMax,arr[idx]);
// }

// bool find(vector<int> &arr, int data, int idx)
// {
//     if (idx == arr.size())
//         return false;

//     if (arr[idx] == data)
//         return true;

//     return find(arr, data, idx + 1);
// }

// void set2()
// {
//     vector<int> arr = {1, 45, 67, 456, 345, 5, 454324, 34, 14, 1, 4312, 4};
//     //10 size ka vector with default int as zero
//     display(arr, 0);
//     cout << endl;
//     cout << maximum(arr, 0) << endl;

//     if (find(arr, 5, 0))

//     {
//         cout << "true";
//     }
//     else
//     {
//         cout << "false";
//     }
// }

vector<string> subseq(string str)
{
    if (str.length() == 0)
    {
        vector<string> base;
        base.push_back(" ");
        return base;
    }

    char ch = str[0];            // pehla character uthaoo  java -- str.charAt(0);
    string nstr = str.substr(1); // str.subString(1);
    vector<string> ans = subseq(nstr);
    vector<string> myans;

    for (string s : ans) // for each loop   - as we are not doing anything  with settig the value
                         //  but we are using it for getting so it will work
    {                    // here we are not worried about the order we are just getting the jo  done
                         // not looking for the formatiing
        myans.push_back(s);
        myans.push_back(ch + s);
    }
    return myans;
}

int printsubseq(string str, string ans)
{
    if (str.length() == 0)
    {
        cout << ans << endl;
        return 1;
    }
    int count = 0;

    count += printsubseq(str.substr(1), ans + str[0]);
    count += printsubseq(str.substr(1), ans);

    return count;
}

void set3()
{
    string str = "abcd";
    vector<string> myans = subseq(str);

    for (string s : myans)
    {
        cout << s << endl;
    }
    cout << endl
         << endl
         << endl
         << endl;

    cout << printsubseq(str, "") << endl;
}

void solve()
{
    set3();
}

int main()
{
    solve();
    return 0;
}